### PR TITLE
test(cmake): exclude test targets from build only configurations

### PR DESCRIPTION
**Pull request by some00**
Related to #2337

### Description of the feature or fix

Possible solution for a build error. Suggested in: https://github.com/lvgl/lvgl/issues/2337#issuecomment-1322689179

### Checkpoints
- [ ] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `LV_ITERATE_ROOTS` list in [`lv_gc.h`](https://github.com/lvgl/lvgl/blob/master/src/misc/lv_gc.h) and mark the variable with `GC_ROOT(variable)` when it's used. See a detaild description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<modul_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the followings needs to be followed (see a detaild description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)): 
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Comment by some00** [comment]
I personally don't recommend this solution. It doesn't scale well, but the source of the problem I think is the usage of globbing for gathering source files. [CMake documentation](https://cmake.org/cmake/help/latest/command/file.html#filesystem) has a note about it.
In short, an already configured build won't detect a newly added file, which can confuse newcomers. A reconfiguration has to be triggered by calling cmake/ccmake or deleting the existing build folders.

I would recommend creating a function out of the current loop and listing the source files. This can be mentioned in the documentation as part of the process to create a new test case.
For the current problem this would mean a simple if statement around the call with `test_meter.c` argument.
The downside is possible conflicts.
@kisvegabor What do you think? Should I change to loop to a function? The `ifdef` in test source file(s) solution would scale better and won't conflict either.

**Comment by kisvegabor** [comment]
Thank you for looking into this issue.

I think all we need in the end is having different source files for build tests (e.g. `OPTIONS_FULL_32BIT`) and different source files for unit tests (e.g. `OPTIONS_TEST_SYSHEAP`). 

- For build tests we need `lvgl/src` + `lvgl/demos` + `lvgl/examples`
- For unit tests we need `lvgl/src` + `lvgl/demos` + `lvgl/tests`

Do you think it's doable?

**Comment by some00** [comment]
I'll update the PR.

**Comment by some00** [comment]
I took the chance to try and reorganize the few hundred lines of compilation options. Hope you won't mind, but again happy to revert that part.

Unfortunately, test cases depend on the examples. One instance is in `test_animimg.c`. Global variable `animimg001` is used, which is defined in `examples/assets/animimg001.c`.
So test cases still depend on `lvgl_examples` target, but that shouldn't be an issue for `test_meter.c`.

**Comment by kisvegabor** [comment]
> Unfortunately, test cases depend on the examples. One instance is in test_animimg.c. Global variable animimg001 is used, which is defined in examples/assets/animimg001.c.

It's not intentional. We have [`tests/src/test_assets`](https://github.com/lvgl/lvgl/tree/master/tests/src/test_assets) and `examples/assets/animimg*.c` should be copied there.

**Comment by kisvegabor** [comment]
Sorry for the trouble, but due to a different issue I moved the `-DLV_...` configs from CMake to header files. See [here](https://github.com/lvgl/lvgl/commit/077cd2e15d103bc71cae8472cf8af9d7dbd1ae87). Could you mind rebasing on top of these changes?

**Comment by some00** [comment]
Is there something I should do about the CI errors? They seem unrelated to my changes.
`Error response from daemon: login attempt to https://ghcr.io/v2/ failed with status: 503 Service Unavailable`

**Comment by kisvegabor** [comment]
It happens time to time. I've restarted the tests.

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -169,7 +181,7 @@ foreach( test_case_fname ${TEST_CASE_FILES} )
         ${test_case_fname}
         ${test_runner_fname}
     )
-    target_link_libraries(${test_name} test_common lvgl_examples lvgl_demos lvgl png m ${TEST_LIBS})
+    target_link_libraries(${test_name} test_common lvgl_demos lvgl png m ${TEST_LIBS})
```
Comment (kisvegabor) [inline tests/CMakeLists.txt:184]: Why was the examples removed from here?

Comment (some00) [reply tests/CMakeLists.txt:184]: Based on your [comment](https://github.com/lvgl/lvgl/pull/3821#issuecomment-1323422245) the unit tests shouldn't build and link to `lvgl_examples`. I must have misunderstood that.

Comment (kisvegabor) [reply tests/CMakeLists.txt:184]: Ah, it's only for the unit test. Got it. Thank you

**Comment by kisvegabor** [comment]
It looks good from my side. Can I merge it?

**Comment by some00** [comment]
Yes, I'm finished.

**Comment by kisvegabor** [comment]
Thank you!

**Pull request merged**